Mechanism for generating proposal ID (#18)
hyperledger-labs/perun-proposals: design/006-proposal-id-generation.md

# Proposal: Mechanism for generating proposal ID

* Author(s): Manoranjith, Matthias Geihs
* Status: propose
* Related issue: https://github.com/hyperledger-labs/go-perun/issues/167 

## Summary

This proposal describes the need for proposal ID and the mechanism for
generating it.

## Motivation

Until now, Proposal ID was being generated by encoding all the parameters of a
proposal using perunio encoding scheme. Recently, an abstraction for wire
serialization was introduced (see
[#233](https://github.com/hyperledger-labs/go-perun/issues/233)), so that
different encoding schemes can be used for serializing wire messages. As a part

> of this abstraction, it was needed to the mechanism for generating proposal ID
> re-evaluate, because it depends on a specific encoding scheme: perunio

Reviewer: ```suggestion
of this abstraction, we needed to re-evaluate the mechanism for generating the proposal ID, because it depended on a specific encoding scheme: perunio
```

Reviewer: Or in passive voice:

"of this abstraction, it was needed to re-evaluate the mechanism for generating the proposal ID, because it depended on a specific encoding scheme: perunio"

Author: Updated in 18b56f0b.

the participants know the nonce share of every other participant. Hence, until
a channel ID can be generated, a unique identifier for identifying the channel
proposals is needed. For this purpose, the proposal ID was introduced.

### Expected properties for a proposal ID

It is sufficient for the proposal ID to be unique for each channel proposal. It
is not necessary to be able to derive the proposal ID from a channel proposal.

### Currently used mechanism

The proposal ID was generated by encoding the channel proposal along with the
nonce share of proposer. Perun protocol allows multiple options for nonce
share, such as using a monotonically increasing counters for each set of
participants or filling a byte array using a cryptographically secure random
number generator. In either case, the participant ensures that, the nonce share
for each proposal between a given set of participants is different. Hence, the
proposal ID was unique for each proposal generated by a participant.

### Issues with currently used mechanism

In [#233](https://github.com/hyperledger-labs/go-perun/issues/233),
abstractions for wire serialization were introduced. After this, it is possible

> to use a different encoding schemes such as protocol buffers. For instance, if

Reviewer: ```suggestion
to use a different encoding scheme such as protocol buffers. For instance, if
```

Author: Updated in this [force-push](https://github.com/hyperledger-labs/perun-proposals/compare/5ac16903f87512ed316835ff43cc5b244448eb50..18b56f0b8584fd616aa1fcec4807b9d79c26767e).

byte array is fixed at 32 bytes.

When using this mechanism, there are three possible scenarios where two channel
proposals can have same ID. In any case, if a user receives multiple channel
proposals with the same ID, they can process only the first one and reject the
latter requests.

1. When a user (say Alice) intentionally tries to send two proposals to another
   user Bob, with the same proposal ID.

   Because Bob receives two proposals with the same proposal ID, they could
   process only the first and reject the second one. Since, this defeats the
   objective of Alice (to establish a channel with Bob), Alice has no incentive
   to do this.

2. By coincidence, if two users generate the same proposal ID.

   When using a cryptographically secure random number generator with
   sufficient entropy, the possibility of this occurring is very rare. Even if
   it occurs due to a rare coincidence, Bob could simply reject the second
   proposal with the same ID.

3. An adversary (say Mallory) tries to intercept the communication between two
   users (say Alice and Bob) and channel proposals with the same ID.

>    TODO

Reviewer: ```suggestion
However, this situation only occurs when mallory uses the same message bus as Alice and Bob or when the connection between is not properly secured (say, for example, using TLS).
```

## Figures

![Sequence of interactions for opening a channel.](006/channel_opening.svg)

**Figure 1**: Sequence of interactions for opening a channel.

## Rationale

Two alternate approaches were evaluated.

1. One approach was to expose interfaces in the wire serialization abstraction,

>    that can encodes a channel proposal using the underlying encoding scheme.

Reviewer: ```suggestion
   that can encode a channel proposal using the underlying encoding scheme.
```

Author: Updated in this [force-push](https://github.com/hyperledger-labs/perun-proposals/compare/5ac16903f87512ed316835ff43cc5b244448eb50..18b56f0b8584fd616aa1fcec4807b9d79c26767e).
